src/music_mix.cpp

```diff
@@ -88,12 +88,16 @@ void KMusic::shutdown_music(void)
  *
  * Sets the volume of the currently playing music.
  *
- * \param   volume 0 (silent) to 100 (loudest)
+ * \param   volume 0 (silent) to 250 (loudest)
  */
-void KMusic::set_music_volume(float volume)
+void KMusic::set_music_volume(int volume)
 {
-    int ivol = int(volume * float(MIX_MAX_VOLUME));
-    Mix_VolumeMusic(ivol);
+    float v = float(volume) / 250.0f;
+    if (v >= 0.0f && v <= 1.0f)
```

Comment:
If the input volume is outside the valid range, would we want to do nothing, or just clamp the inputs?

`std::clamp()` from `<algorithm>` or `SDL_clamp()` from `SDL_stdinc.h` (probably included implicitly) could be used.

```c++
void KMusic::set_music_volume(int volume)
{
    constexpr float MaxMusicVolume = 250.0f; // or whatever will give a ratio within [0.0, 1.0]
    int ivol = std::clamp(int(float(MIX_MAX_VOLUME * volume) / MaxMusicVolume), 0, MIX_MAX_VOLUME);
    Mix_VolumeMusic(ivol);
}
```

I ran through some scenarios both with the old `[0,100]` range:
```
  inputs: -3, -2,  -1          ; outputs:   0,   0,   0
  inputs:  0,  1,   2,   3     ; outputs:   0,   1,   2,   3
  inputs: 98, 99, 100, 101, 102; outputs: 125, 126, 128, 128, 128
```
and with the new `[0,250]` range:
```
  inputs:  -3,  -2,  -1          ; outputs:   0,   0,   0
  inputs:   0,   1,   2,   3     ; outputs:   0,   0,   1,   1
  inputs:  98,  99, 100, 101, 102; outputs:  50,  50,  51,  51,  52
  inputs: 126, 127, 128, 129, 130; outputs:  64,  65,  65,  66,  66
  inputs: 248, 249, 250, 251, 252; outputs: 126, 127, 128, 128, 128
```

So all the clamping works as desired, if this is what we want. Some notes on the range change, though:

- When using the 0..100 range, the lower volumes like 1, 2 and 3 and upper volumes like 126, 127 and 128 increased the volume on each step (not a direct 1:1 correlation, but probably close enough).
- When using the 0..250 range, the input needs to be changed quite a bit more before we see changes, like seen above.

Reply:
Looking at the old code I'm not sure there was ever a 0..100 range, despite the comment. The music volume seems to be passed in as a float 0.0..1.0, and it was derived by dividing by 250. e.g.
https://github.com/OnlineCop/kq-fork/blob/de7915f09b77bec9083d65334cff6793d0f70d63/src/setup.cpp#L444
It makes sense to me have 0..250 for both overall and music volumes when shown in the UI and saved to the config file. Then convert internally to whatever SDL / SDL_mixer require.
On clamping - possibly this was a holdover from Allegro - it had one command taking two params to set both digital audio and midi volumes, and you would pass -1 to say "don't change this one"
Clamp makes more sense.